pkg/controllers/user/networkpolicy/clusterNetPolHandler.go

```diff
@@ -0,0 +1,102 @@
+package networkpolicy
+
+import (
+	"fmt"
+
+	"github.com/rancher/rancher/pkg/controllers/user/nodesyncer"
+	"github.com/rancher/types/apis/core/v1"
+	"github.com/rancher/types/apis/management.cattle.io/v3"
+	"github.com/rancher/types/config"
+	"github.com/sirupsen/logrus"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+)
+
+type clusterNetPolHandler struct {
+	cluster       *config.UserContext
+	pnpLister     v3.ProjectNetworkPolicyLister
+	podLister     v1.PodLister
+	serviceLister v1.ServiceLister
+	projClient    v3.ProjectInterface
+	clusters      v3.ClusterInterface
+	npmgr         *netpolMgr
+}
+
+func (ch *clusterNetPolHandler) Sync(key string, cluster *v3.Cluster) error {
+	if cluster == nil || cluster.DeletionTimestamp != nil ||
+		!v3.ClusterConditionReady.IsTrue(cluster) ||
+		cluster.Status.AppliedEnableNetworkPolicy == cluster.Spec.EnableNetworkPolicy {
+		return nil
+	}
+
+	cluster.Status.AppliedEnableNetworkPolicy = cluster.Spec.EnableNetworkPolicy
+	if _, err := ch.clusters.Update(cluster); err != nil {
+		return err
+	}
+
+	if err := ch.refresh(cluster); err != nil {
+		// reset if failure
+		cluster.Status.AppliedEnableNetworkPolicy = !cluster.Spec.EnableNetworkPolicy
+		if _, err := ch.clusters.Update(cluster); err != nil {
+			return err
+		}
+		return err
+	}
+
+	return nil
+}
+
+func (ch *clusterNetPolHandler) refresh(cluster *v3.Cluster) error {
+	if cluster.Status.AppliedEnableNetworkPolicy {
+		logrus.Debugf("clusterNetPolHandler: calling sync to create network policies")
+		projects, err := ch.npmgr.projLister.List(cluster.Name, labels.NewSelector())
+		if err != nil {
+			return err
+		}
+		for _, project := range projects {
+			ch.projClient.Controller().Enqueue(project.Namespace, project.Name)
+		}
+
+		pods, err := ch.podLister.List("", labels.NewSelector())
+		if err != nil {
+			return err
+		}
+		for _, pod := range pods {
+			ch.cluster.Core.Pods("").Controller().Enqueue(pod.Namespace, pod.Name)
```

Comment:
trigger for pods which have hostports, not for all. You can pull out the code to find the pods which have host ports from the other function to a utils function.